### PR TITLE
Initial indentation parsing #70

### PR DESCRIPTION
**Pull request by g4s8**
Adapted Python3 parser for #70 issue

### CODE REVIEW

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 [hunk 1]
```diff
@@ -78,12 +155,11 @@ arguments_declaration returns [List<Argument> ret]
 
 argument_declaration returns [Argument ret]
     :
-    HINAME
-    SPACE
-    LONAME
+    HINAME SPACE LONAME
```
Comment (yegor256) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: @g4s8 why did you change the formatting? I think we should keep tokens one per line, like everywhere else in this file.

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 [hunk 2]
```diff
@@ -68,7 +145,7 @@ arguments_declaration returns [List<Argument> ret]
         head=argument_declaration
         { $ret.add($head.ret); }
         (
-            COMMA
+            COMMA SPACE
```
Comment (yegor256) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: @g4s8 same here, let's use "one token per line" rule

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 [hunk 3]
```diff
@@ -9,16 +9,109 @@ grammar Program;
...
         type_declaration
         { types.add($type_declaration.ret); }
-        |
-        object_instantiation
-        |
-        object_copying
+//        |
+//        object_instantiation
+//        |
+//        object_copying
```
Comment (yegor256) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: @g4s8 don't comment anything, just delete. This is what Git is for - to make it possible to get back in history.

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 [hunk 4]
```diff
@@ -9,16 +9,109 @@ grammar Program;
...
     :
     { Collection<Type> types = new LinkedList<Type>(); }
     (
         type_declaration
         { types.add($type_declaration.ret); }
-        |
-        object_instantiation
-        |
-        object_copying
+//        |
```
Comment (yegor256) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: @g4s8 looks like you're deleting a very big piece of functionality. Why? You need to add a `@todo` [puzzle](https://github.com/yegor256/pdd), with an explanation or keep this functionality in place (a better option).

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 [hunk 5]
```diff
@@ -27,36 +120,20 @@ program returns [Tree ret]
 type_declaration returns [Type ret]
     :
     { Collection<Method> methods = new LinkedList<Method>(); }
-    TYPE
-    SPACE
-    HINAME
+    TYPE SPACE HINAME (SPACE EXTENDS SPACE HINAME (COMMA SPACE HINAME)*)? COLON
+    NEWLINE INDENT
```
Comment (yegor256) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: @g4s8 let's keep them in multiple lines

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 [hunk 6]
```diff
@@ -27,36 +120,20 @@ program returns [Tree ret]
...
     { $ret = new Type($HINAME.text, methods); }
     ;
 
 method_declaration returns [Method ret]
     :
-    HINAME
-    SPACE
-    LONAME
-    arguments_declaration
+    HINAME SPACE LONAME arguments_declaration
```
Comment (yegor256) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: @g4s8 same here, multiple lines please

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 [hunk 7]
```diff
@@ -199,5 +277,40 @@ NUMBER: ( '0' .. '9' )+;
...
 
-SPACE: ' ';
-EOL: '\n';
+fragment SPACES: [ \t]+;
+
+NEWLINE
+ : ( {atStartOfInput()}?   SPACES
+   | ( '\r'? '\n' | '\r' ) SPACES?
+   )
+   {
```
Comment (yegor256) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: @g4s8 I realize that you got this code somewhere, right? In that case, we have to put a link to the original source.

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 [hunk 8]
```diff
@@ -199,5 +277,40 @@ NUMBER: ( '0' .. '9' )+;
...
+     String newLine = getText().replaceAll("[^\r\n]+", "");
+     String spaces = getText().replaceAll("[\r\n]+", "");
+     int next = _input.LA(1);
+     if (opened > 0 || next == '\r' || next == '\n' || next == '#') {
+       // If we're inside a list or on a blank line, ignore all indents,
+       // dedents and line breaks.
+       skip();
+     }
```
Comment (yegor256) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: @g4s8 I would try to preserve our original code formatting here:

```
if () {
} else {
}
```

**Comment by g4s8** [comment]
@yegor256 thanks for review! I've corrected comments and pushed changes.

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 (hunk 4 above)
Comment (g4s8) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: I've kept object parsing rules with addition of `INDENT` and `DEDENT` tokens. Also I've put `@todo` puzzle into ignored object test.

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 (hunk 5 above)
Comment (g4s8) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: Fixed.

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 (hunk 6 above)
Comment (g4s8) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: Fixed.

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 (hunk 2 above)
Comment (g4s8) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: Fixed.

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 (hunk 7 above)
Comment (g4s8) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: Added link as comment.

File: eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4 (hunk 8 above)
Comment (g4s8) [inline eo-compiler/src/main/antlr4/org/eolang/compiler/Program.g4]: Fixed.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by g4s8** [comment]
> ERROR: "puzzle eo-compiler/src/test/java/org/eolang/compiler/ProgramTest.java:48-48 has a very short description of just 11 words while a minimum of 20 is required"
> ERROR: "puzzle eo-compiler/src/test/java/org/eolang/compiler/ProgramTest.java:48-48 has an estimate of 0 minutes, which is lower than 15 minutes"

@yegor256 merge failed with PDD error caused by incorrect syntax of `@todo`. I've fixed rultor errors, coluld you please restart him?

**Comment by yegor256** [comment]
@rultor try to merge again

**Comment by yegor256** [comment]
@g4s8 many thanks!
